crates/plugins/src/install.rs

```diff
@@ -0,0 +1,288 @@
+                    git_source.pull().await?;
+                }
+                let file = File::open(
+                    &self
+                        .plugins_dir
+                        .join(PLUGINS_REPO_LOCAL_DIRECTORY)
+                        .join(PLUGINS_REPO_MANIFESTS_DIRECTORY)
+                        .join(&info.name)
+                        .join(get_manifest_file_name_version(&info.name, &info.version)),
+                )
+                .map_err(|_| {
+                    anyhow!(
+                        "Could not find plugin [{} {:?}] in centralized repository",
+                        info.name,
+                        info.version
+                            .as_ref()
+                            .map(|s| s.to_string())
+                            .unwrap_or_else(|| String::from("latest"))
+                    )
+                })?;
+                serde_json::from_reader(file)?
+            }
+        };
+
+        // Disallow installing plugins with the same name as spin internal
+        // subcommands
+        if SPIN_INTERNAL_COMMANDS
+            .iter()
+            .any(|&s| s == plugin_manifest.name())
+        {
+            bail!(
+                "Trying to install a plugin with the same name '{}' as an internal plugin",
+                plugin_manifest.name()
+            );
+        }
+
+        // Disallow downgrades and reinstalling identical plugins
+        if let Ok(installed) = get_plugin_manifest(&plugin_manifest.name(), &self.plugins_dir) {
+            if installed.version > plugin_manifest.version || installed == plugin_manifest {
+                bail!(
+                    "plugin {} already installed with version {} but attempting to install same or older version ({})",
+                    installed.name(),
+                    installed.version,
+                    plugin_manifest.version,
+                );
+            }
+        }
+
+        assert_supported_version(&self.spin_version, &plugin_manifest.spin_compatibility)?;
+
+        let os: Os = if cfg!(target_os = "windows") {
+            Os::Windows
+        } else if cfg!(target_os = "linux") {
+            Os::Linux
+        } else if cfg!(target_os = "macos") {
+            Os::Osx
+        } else {
+            bail!("This plugin is not supported on this OS");
+        };
+        let arch = std::env::consts::ARCH;
+        let plugin_package = plugin_manifest
+            .packages
+            .iter()
+            .find(|p| p.os == os && p.arch.to_string() == arch)
+            .ok_or_else(|| anyhow!("This plugin does not support this OS or architecture"))?;
+        let target = plugin_package.url.to_owned();
+
+        // Ask for user confirmation to install if not overridden with CLI
+        // option
+        if !self.yes_to_all
+            && !Prompter::new(&plugin_manifest.name(), &plugin_manifest.license, &target)?.run()?
+        {
+            // User has requested to not install package, returning early
```

Comment:
If the user says no, and this was a downgrade, will this result in the user's existing plugin being deleted rather being a no-op?

Reply:
Good point. I fixed this by removing an uninstall for downgrade scenarios and just installing over it.